src/VisualStudio/IntegrationTest/TestUtilities/InProcess/Debugger_InProc.cs

```diff
@@ -0,0 +1,130 @@
+{
+    internal class Debugger_InProc : InProcComponent
+    {
+        public static Debugger_InProc Create() => new Debugger_InProc();
+
+        public void StartDebugging(bool waitForBreakMode)
+        {
+            var debuggerEventsMonitor = new VsDebuggerEventsMonitor(GetDTE().Events.DebuggerEvents);
+            if (waitForBreakMode)
+            {
+                StartAndWaitForDebuggerBreakMode(debuggerEventsMonitor);
+            }
+            else
+            {
+                StartAndWaitForDebuggerRunMode(debuggerEventsMonitor);
+            }
+        }
+
+        public void Continue(bool waitForBreakMode)
+        {
+            // Sometimes you will get a COM exception if you attempt to continue too quickly
+            RetryHelper.Try(() => GetDTE().Debugger.Go(waitForBreakMode));
+        }
+
+        private void StartAndWaitForDebuggerDesignMode(VsDebuggerEventsMonitor debuggerEventsMonitor)
+        {
+            using (var semaphore = new SemaphoreSlim(1))
+            {
+                semaphore.Wait();
+                VsDebuggerEventsMonitor.DebuggerModeChangeEventHandler releaseSemaphore = (mode, reason) => semaphore.Release();
+                debuggerEventsMonitor.OnEnterDesignMode += releaseSemaphore;
+                ExecuteCommand("Debug.Start");
+                semaphore.Wait();
+                debuggerEventsMonitor.OnEnterDesignMode -= releaseSemaphore;
+            }
+        }
+
+        private void StartAndWaitForDebuggerRunMode(VsDebuggerEventsMonitor debuggerEventsMonitor)
+        {
+            using (var semaphore = new SemaphoreSlim(1))
+            {
+                semaphore.Wait();
+                VsDebuggerEventsMonitor.DebuggerModeChangeEventHandler releaseSemaphore = (mode, reason) => semaphore.Release();
+                debuggerEventsMonitor.OnEnterRunMode += releaseSemaphore;
+                ExecuteCommand("Debug.Start");
+                semaphore.Wait();
+                debuggerEventsMonitor.OnEnterDesignMode -= releaseSemaphore;
+            }
+        }
+
+        private void StartAndWaitForDebuggerBreakMode(VsDebuggerEventsMonitor debuggerEventsMonitor)
+        {
+            using (var semaphore = new SemaphoreSlim(1))
+            {
+                semaphore.Wait();
+                VsDebuggerEventsMonitor.DebuggerModeChangeEventHandler releaseSemaphore = (mode, reason) => semaphore.Release();
+                debuggerEventsMonitor.OnEnterBreakMode += releaseSemaphore;
+                ExecuteCommand("Debug.Start");
+                semaphore.Wait();
+                debuggerEventsMonitor.OnEnterBreakMode -= releaseSemaphore;
+            }
+        }
+
+        public enum DebuggerMode
+        {
+            BreakMode = dbgDebugMode.dbgBreakMode,
+            DesignMode = dbgDebugMode.dbgDesignMode,
+            RunMode = dbgDebugMode.dbgRunMode
+        }
+
+        internal class VsDebuggerEventsMonitor
```

Comment:
Consider: if you're going to have this always subscribed, then just have this maintain a set of ManualResetEvents that the other helpers can just .Wait() on rather than having the extra event subscription stuff going on with the semaphores.